feat(AppFrame): Navigation top bar and alerts

ashbork opened the pull request:
## Description
Adds two components: `NavigationTopBar` and `NavigationTopBarAlert` (`NavigationTopBar.Alert`). Allows for adding alerts of different kinds with sensible defaults following the DS guidelines.

## Storybook

https://feature-navigation-top-bar--613a8e945a5665003a05113b.chromatic.com

## Checklist

**Obligatory:**

- [x] Self review (use this as your final check for proposed changes before requesting the review)
- [x] Add correct label
- [ ] Assign pull request with the correct issue

ashbork committed:
feat(AppFrame): add NavTopBar with Alerts

ashbork committed:
feat(AppFrame): add documentation for NavTopBar

ashbork committed:
feat(AppFrame): add examples of NavTopBar usage

ashbork commented on packages/react-components/src/components/AppFrame/components/NavigationTopBar/NavigationTopBar.module.scss:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    &__wrapper {
+      // Grid trick: https://css-tricks.com/css-grid-can-do-auto-height-transitions/
```
This may be confusing - recommend reading before asking abt the grid wrapper

ashbork commented on packages/react-components/src/components/AppFrame/components/NavigationTopBar/NavigationTopBar.module.scss:
```diff
@@ -0,0 +1,112 @@
+      gap: var(--spacing-4);
+      align-items: center;
+      justify-content: center;
+      width: calc(100% - 48px);
```
places the 'x' in a sensible location

ashbork commented on packages/react-components/src/components/AppFrame/components/NavigationTopBar/NavigationTopBar.module.scss:
```diff
@@ -0,0 +1,112 @@
+      }
+    }
+
+    &__close-button {
```
could use a DS button instead, but none of them are quite what we want - an inert icon btn with no hover animation

ashbork commented on packages/react-components/src/components/AppFrame/components/NavigationTopBar/NavigationTopBar.tsx:
```diff
@@ -0,0 +1,140 @@
+  show = true,
+}) => {
+  const alertRef = React.useRef<HTMLDivElement>(null);
+  const { isMounted, isOpen } = useAppFrameAnimations({
```
While this hook isn't specifically designed for this component, it does work

marcinsawicki replied:
This hook is designed for `AppFrame`, and top bar is a part of the whole component, so I don't see any problems with using it here :)

ashbork commented on packages/react-components/src/components/AppFrame/components/NavigationTopBar/NavigationTopBar.tsx:
```diff
@@ -0,0 +1,140 @@
+    elementRef: alertRef,
+  });
+
+  const isMobile = window.matchMedia('(max-width: 768px)').matches;
```
There should be a better way, but I didn't find anything that we're using in DS currently - the check only updates on rerenders which isn't great, we don't want to `useEffect` either. CSS feels janky for this purpose. Thoughts?

marcinsawicki replied:
Why do we need to change the kind of button for mobile only? 🤔

ashbork replied:
That's how it is in the design, presumably because the secondary button takes up much less space

ashbork committed:
fix(NavigationTopBar): fix padding for consistent sizing

ashbork committed:
fix(AppFrame): remove unnecessary transition

ashbork committed:
fix(AppFrame): add optional `title` prop to NavigationTopBar
this allows for a title (or other arbitrary content) to be added under the alerts

vladko-uxds reviewed (approved):
👍

VadymBezpalko commented on packages/react-components/src/components/AppFrame/AppFrame.mdx:
````diff
@@ -160,6 +160,74 @@ const SideNavigationSwitcher: FC = () => {
+```
+
+
+### Top bar alert <a id="TopBarAlert" />
````
We can include this section in "How to build the navigation" navigation links (line 20)

VadymBezpalko commented on packages/react-components/src/components/AppFrame/components/NavigationTopBar/NavigationTopBar.module.scss:
```diff
@@ -0,0 +1,119 @@
+      justify-content: center;
+      width: calc(100% - 48px);
+
+      @media screen and (width <= 768px) {
```
we can store width value in the variable and reuse it in the code below

marcinsawicki commented on packages/react-components/src/components/AppFrame/AppFrame.mdx:
```diff
@@ -160,6 +160,74 @@ const SideNavigationSwitcher: FC = () => {
+<AppFrame
+  navigation={<Navigation>...</Navigation>}
+>
+  <NavigationTopBar>
+    <NavigationTopBarAlert kind="info">
+    Hello, this is an alert!
+    </NavigationTopBarAlert>
+  </NavigationTopBar>
```
Shouldn't this be placed in the `topBar` prop in this example? 🤔

marcinsawicki commented:
First of all, great job with adding the new sub-components for the `AppFrame` and improving how the top bar works. 🙌 
Left some comments.
Missing unit tests.

VadymBezpalko commented on packages/react-components/src/components/AppFrame/components/NavigationTopBar/types.ts:
```diff
@@ -0,0 +1,61 @@
+  /**
+   * Show or hide the alert, defaults to `true`. Changes to this prop are animated - the alert will enter and leave smoothly.
+   * */
+  show?: boolean;
```
Historically, our codebase commonly uses `isVisible` naming. Therefore, I suggest using similar naming to maintain a consistent pattern.
P.S.: Consistent naming patterns are something I highly desire, but for various reasons, we don't have them yet. 😄

ashbork replied:
> but for various reasons, we don't have them yet. 😄

sorry to have been one of these reasons! it's `isVisible` now

VadymBezpalko replied:
@ashbork Haha, no, I meant legacy code from the previous DS version, since part of the new API is a clone of the old one. 😄 
You and the other contributors are actually the reason why the codebase keeps improving, so great job and thanks for your effort! 🏆

ashbork committed:
fix(AppFrame): improve naming consistency

ashbork committed:
fix(AppFrame): extract mobile breakpoint variable

ashbork committed:
feat(AppFrame): change additional nodes prop name, add title component

ashbork committed:
fix(AppFrame): add aria role

This gives us an added benefit of the top bar becoming a polite live region,
announcing changes on the bar. (`alert` is too intrusive IMHO)

ashbork committed:
fix(AppFrame): make children optional
no need to force the component to have children

ashbork committed:
test(AppFrame): add unit tests for NavTopBar

ashbork committed:
docs(AppFrame): improve documentation

ashbork committed:
Merge branch 'main' into feature-navigation-top-bar

VadymBezpalko committed:
Merge branch 'main' into feature-navigation-top-bar
